integration_tests/src/main/python/hash_aggregate_test.py

```diff
@@ -510,41 +488,47 @@ def test_hash_multiple_grpby_pivot(data_gen, conf):
 
 @approximate_float
 @ignore_order(local=True)
-@allow_non_gpu('HashAggregateExec', 'PivotFirst', 'AggregateExpression', 'Alias', 'GetArrayItem',
-        'Literal', 'ShuffleExchangeExec')
 @incompat
-@pytest.mark.parametrize('data_gen', [_grpkey_floats_with_nulls_and_nans], ids=idfn)
-def test_hash_pivot_reduction_nan_fallback(data_gen):
-    # collect values to pivot in previous, to avoid this preparation job being captured
-    def fetch_pivot_values(spark):
-        max_values = spark._jsparkSession.sessionState().conf().dataFramePivotMaxValues()
-        df = gen_df(spark, data_gen, length=100)
-        df = df.select('b').distinct().limit(max_values + 1).sort('b')
-        return [row[0] for row in df.collect()]
-
-    pivot_values = with_cpu_session(fetch_pivot_values)
-
-    assert_gpu_fallback_collect(
+@pytest.mark.parametrize('data_gen', _init_list_no_nans, ids=idfn)
+@pytest.mark.parametrize('conf', get_params(_confs_with_nans, params_markers_for_confs_nans), ids=idfn)
+def test_hash_reduction_pivot_without_nans(data_gen, conf):
+    assert_gpu_and_cpu_are_equal_collect(
         lambda spark: gen_df(spark, data_gen, length=100)
             .groupby()
-            .pivot('b', pivot_values)
+            .pivot('b')
             .agg(f.sum('c')),
-        "PivotFirst",
-        conf=_nans_float_conf)
+        conf=conf)
 
 @approximate_float
 @ignore_order(local=True)
 @incompat
-@pytest.mark.parametrize('data_gen', _init_list_no_nans, ids=idfn)
+@pytest.mark.parametrize('data_gen', _init_list_with_nans_and_no_nans, ids=idfn)
 @pytest.mark.parametrize('conf', get_params(_confs_with_nans, params_markers_for_confs_nans), ids=idfn)
-def test_hash_reduction_pivot_without_nans(data_gen, conf):
+def test_hash_reduction_pivot_with_nans(data_gen, conf):
     assert_gpu_and_cpu_are_equal_collect(
         lambda spark: gen_df(spark, data_gen, length=100)
             .groupby()
             .pivot('b')
             .agg(f.sum('c')),
         conf=conf)
 
+@approximate_float
+@ignore_order(local=True)
+@allow_non_gpu('HashAggregateExec', 'PivotFirst', 'AggregateExpression', 'Alias', 'GetArrayItem',
+        'Literal', 'ShuffleExchangeExec', 'HashPartitioning', 'KnownFloatingPointNormalized',
+        'NormalizeNaNAndZero')
+@incompat
+@pytest.mark.parametrize('data_gen', [_grpkey_floats_with_nulls_and_nans], ids=idfn)
+def test_hash_pivot_groupby_duplicates_fallback(data_gen):
```

Comment:
Thanks for adding this test. I had documented in GpuOverrides but forgot to add the test when pivot values has duplicates.